pkg/network/ebpf/c/protocols/postgres/decoding.h

```diff
@@ -240,20 +283,33 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
         // the message tag. So we need to add 1 to the message length to jump over the entire message.
         pktbuf_advance(pkt, header.message_len + 1);
     }
+    iteration_value->total_msg_count += messages_count;
 
     if (found_command_complete) {
+        handle_command_complete(&conn_tuple, transaction);
+        update_msg_count_telemetry(pg_msg_counts, iteration_value->total_msg_count);
+
         return 0;
     }
-    // We didn't find a command complete message, so we need to continue processing the packet.
-    // We save the current data offset and increment the iteration counter.
-    iteration_value->iteration += 1;
-    iteration_value->data_off = pktbuf_data_offset(pkt);
 
-    // If the maximum number of tail calls has been reached, we can skip invoking the next tail call.
-    if (iteration_value->iteration >= POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES) {
+    if (iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)) {
+        // reached max messages, add counter and stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->reached_max_messages, 1);
+    }
+    if (pktbuf_data_offset(pkt) == pktbuf_data_end(pkt)) {
+        // stop the iterator if the end of the TCP packet is reached.
+        return 0;
+    }
+    if (read_result == false) {
+        // the packet was fragmented, add counter stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->fragmented_packets, 1);
         return 0;
     }
 
+    // We didn't find a command complete message, so we need to continue processing the packet.
```

Comment:
if `iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)` you're performing additional tail call (that will bail out).
But each tail call has its performance impact, so if you don't need to trigger a new tail call - don't